fabric-api-base/src/main/java/net/fabricmc/fabric/impl/base/event/ArrayBackedEvent.java

```diff
@@ -43,16 +61,150 @@ void update() {
 
 	@Override
 	public void register(T listener) {
+		register(DEFAULT_PHASE, listener);
+	}
+
+	@Override
+	public void register(Identifier phaseIdentifier, T listener) {
+		Objects.requireNonNull(phaseIdentifier, "Tried to register a listener for a null phase!");
 		Objects.requireNonNull(listener, "Tried to register a null listener!");
 
-		lock.lock();
+		synchronized (lock) {
+			getOrCreatePhase(phaseIdentifier).addListener(listener);
+			rebuildInvoker(handlers.length + 1);
+		}
+	}
+
+	private EventPhaseData<T> getOrCreatePhase(Identifier id) {
+		EventPhaseData<T> phase = phases.get(id);
+
+		if (phase == null) {
+			phase = new EventPhaseData<>(id, handlers.getClass().getComponentType());
+			phases.put(id, phase);
+			sortedPhases.add(phase);
+		}
+
+		return phase;
+	}
+
+	private void rebuildInvoker(int newLength) {
+		// Rebuild handlers.
+		if (sortedPhases.size() == 1) {
+			// Special case with a single phase: use the array of the phase directly.
+			handlers = sortedPhases.get(0).listeners;
+		} else {
+			@SuppressWarnings("unchecked")
+			T[] newHandlers = (T[]) Array.newInstance(handlers.getClass().getComponentType(), newLength);
+			int newHandlersIndex = 0;
+
+			for (EventPhaseData<T> existingPhase : sortedPhases) {
+				for (T handler : existingPhase.listeners) {
+					newHandlers[newHandlersIndex++] = handler;
+				}
+			}
+
+			handlers = newHandlers;
+		}
+
+		// Rebuild invoker.
+		update();
+	}
+
+	@Override
+	public void addPhaseOrdering(Identifier firstPhase, Identifier secondPhase) {
+		Objects.requireNonNull(firstPhase, "Tried to add an ordering for a null phase.");
+		Objects.requireNonNull(secondPhase, "Tried to add an ordering for a null phase.");
+		if (firstPhase.equals(secondPhase)) throw new IllegalArgumentException("Tried to add a phase that depends on itself.");
+
+		synchronized (lock) {
+			EventPhaseData<T> first = getOrCreatePhase(firstPhase);
+			EventPhaseData<T> second = getOrCreatePhase(secondPhase);
+			first.subsequentPhases.add(second);
+			second.previousPhases.add(first);
+			sortPhases();
```

Comment:
Given these sorts are quite expensive, can we have an option to add multiple orderings at once (so the sort is only done after adding all of those ordering definitions)

Reply:
If P is the number of phases and C is the number of constraints, the sort is only O(P+C), so it shouldn't be a big bottleneck. Event registrations themselves are O(N) each after all.

Reply:
But if we try to register C constraints, the current model has a total time complexity of O(C(P+C)) as you do such a calculation after registering each constraint. Often you would register multiple constraints at once, such as when declaring the default phases of an event.

Reply:
Imo if this is a bottleneck then we should provide a way to register multiple constraints at once. But otherwise I'd like to keep the API changes minimal.